### ISSUE TITLE
Freeze application that uses PyInstaller in Python code

### ISSUE THREAD

**Issue by masoudr** [pyinstaller/pyinstaller#4808]
## Description of the issue
In my application, I'm using pyinstaller to freeze my code as described in [here](https://pyinstaller.readthedocs.io/en/stable/usage.html#running-pyinstaller-from-python-code) .
This works fine but after freezing the code I get this error:

> PyInstaller cannot check for assembly dependencies.
> Please install pywin32-ctypes.
> 
> pip install pywin32-ctypes

And that is the cause of using pyinstaller inside my script.
It seems that pyinstaller can't handle the hook but I would appreciate it if you give me a workaround for this.

## Context information (for bug reports)

* Output of `pyinstaller --version`: ```3.6+6d4cce1f8e```
* Version of Python: 3.7
* Platform: Windows (EN)

### A minimal example program which shows the error

```
import os
import PyInstaller.__main__

package_name = 'mypackage'

PyInstaller.__main__.run([
    '--name=%s' % package_name,
    os.path.join('my_package', '__main__.py'),
])

```

### Stacktrace / full error message

```
PyInstaller cannot check for assembly dependencies.
Please install pywin32-ctypes.

pip install pywin32-ctypes
```

**Comment by Legorooj** [comment]
@masoudr can you:

* Install the latest development version? You're out of date.

* Try `pip install pywin32-ctypes`?

**Comment by masoudr** [comment]
@Legorooj I tried the latest development version (`4.0.dev0+b3dd91c8a8`) and the same thing happened. I also have the `pywin32-ctypes` in my packages.

**Comment by htgoebel** [comment]
Did I understand this correctly: Your application uses PyInstaller (as shown above), then you freeze this application and you get the above error? Then obviously some packages from pywin32-ctypes are not collected. Please have a look at [Make sure everything is packaged correctly](https://github.com/pyinstaller/pyinstaller/wiki/How-to-Report-Bugs#make-sure-everything-is-packaged-correctly) for tips to find out what is missing.

**Comment by masoudr** [comment]
@htgoebel You are right. I have the pywin32 packages (I have `pypiwin32`). As soon as I remove the pyinstaller's import the error would be gone. How can I collect the missing package? (I've already used `hidden-import`). I also tried the `pyi-archive_viewer` and I have all the pywin32 packages inside my environment. I think it is a known problem when trying to freeze a package that uses PyInstaller.

**Comment by Legorooj** [comment]
@masoudr are you definitely running from the same environment as pyinstaller has been installed in?

**Comment by masoudr** [comment]
@Legorooj I am. Currently, I ended up creating a new environment with PyInstaller installed and bring the whole directory as a DATA folder.

**Comment by Legorooj** [comment]
@masoudr hang on. You're trying to use PyInstaller from within a PyInstaller executable?

**Comment by masoudr** [comment]
@Legorooj Yes. I'm using PyInstaller inside my code.

**Comment by Legorooj** [comment]
@masoudr have you tried writing a hook for PyInstaller? `collect_all`.

**Comment by masoudr** [comment]
@Legorooj Thanks for the help, but I'm not quite sure how to write the hook, I've looked into the docs about how to create a hook but I don't know which `datas` or `binaries` should I use with `collect_all`.  On the other hand, the error is quite strange for me as it is complaining about missing an existing module which I don't think is related to hooking the PyInstaller itself.

**Comment by Legorooj** [comment]
```
- script_to_compile.py
- hooks
  - hook-PyInstaller.py
```
And inside the hook:
```
from PyInstaller.utils.hooks import collect_all
datas, binaries, hiddenimports = collect_all('PyInstaller')
```
And `--additional-hooks-dir=hooks doesn't work?

> On the other hand, the error is quite strange for me as it is complaining about missing an existing module which I don't think is related to hooking the PyInstaller itself.

PyInstaller needs `pywin32-ctypes` to build scripts.

**Comment by Legorooj** [comment]
@masoudr is this working yet?

**Comment by masoudr** [comment]
@Legorooj I apologize for the late response.
I tried what you did; created a test script:
```
import PyInstaller.__main__

input("It works!")

```
And put the `hook` directory and the contents next to it and froze the app. The output file size got larger but I got the same error.

**Comment by Legorooj** [comment]
@masoudr did you use the `--additional-hooks-dir=hooks` option? It won't work otherwise

**Comment by masoudr** [comment]
@Legorooj Yes, I did, and as I said I checked the logs and the hook worked but got the same output.

**Comment by Legorooj** [comment]
@masoudr I'll try and replicate.

**Comment by Legorooj** [comment]
@masoudr sorry it took me so long to get round to this.

So this isn't a trivial problem to fix. I managed to replicate you're problem with the following hook files inside the `hooks` dir:

#### `hook-altgraph.py`
```python
from PyInstaller.utils.hooks import copy_metadata

datas = copy_metadata('altgraph')
```
#### `hook-PyInstaller.py`
```python
from PyInstaller.utils.hooks import collect_all

datas, binaries, hiddenimports = collect_all('PyInstaller')

hiddenimports += [
    'win32ctypes'
]
```
#### `hook-win32ctypes.py`
```
from PyInstaller.utils.hooks import collect_all

datas, binaries, hiddenimports = collect_all('win32ctypes')
```
After I'd built with those hooks, I got the next error:
```
(.venv) D:\PyInstaller-Issues>dist\test\test.exe
Hello, world
109 INFO: PyInstaller: 4.0.dev0+gbcf7faee.mod
109 INFO: Python: 3.6.8
112 INFO: Platform: Windows-10-10.0.18362-SP0
127 INFO: wrote D:\PyInstaller-Issues\uac_admin_test.spec
130 INFO: UPX is not available.
133 INFO: Extending PYTHONPATH with paths
['D:\\PyInstaller-Issues', 'D:\\PyInstaller-Issues']
146 INFO: checking Analysis
187 INFO: checking PYZ
213 INFO: checking PKG
215 INFO: Building because toc changed
216 INFO: Building PKG (CArchive) PKG-00.pkg
[17096] Failed to execute script test
Traceback (most recent call last):
  File "test.py", line 8, in <module>
    'D:\\PyInstaller-Issues\\uac_admin_test.py', '--onefile'
  File "lib\site-packages\pyinstaller-4.0.dev0+gbcf7faee.mod-py3.6.egg\PyInstaller\__main__.py", line 114, in run
  File "lib\site-packages\pyinstaller-4.0.dev0+gbcf7faee.mod-py3.6.egg\PyInstaller\__main__.py", line 65, in run_build
  File "lib\site-packages\pyinstaller-4.0.dev0+gbcf7faee.mod-py3.6.egg\PyInstaller\building\build_main.py", line 711, in main
  File "lib\site-packages\pyinstaller-4.0.dev0+gbcf7faee.mod-py3.6.egg\PyInstaller\building\build_main.py", line 658, in build
  File "D:\PyInstaller-Issues\uac_admin_test.spec", line 33, in <module>
    console=True )
  File "lib\site-packages\pyinstaller-4.0.dev0+gbcf7faee.mod-py3.6.egg\PyInstaller\building\api.py", line 436, in __init__
  File "lib\site-packages\pyinstaller-4.0.dev0+gbcf7faee.mod-py3.6.egg\PyInstaller\building\api.py", line 199, in __init__
  File "lib\site-packages\pyinstaller-4.0.dev0+gbcf7faee.mod-py3.6.egg\PyInstaller\building\datastruct.py", line 160, in __postinit__
  File "lib\site-packages\pyinstaller-4.0.dev0+gbcf7faee.mod-py3.6.egg\PyInstaller\building\api.py", line 276, in assemble
  File "lib\site-packages\pyinstaller-4.0.dev0+gbcf7faee.mod-py3.6.egg\PyInstaller\depend\bindepend.py", line 948, in get_python_library_path
OSError: Python library not found: libpython36.dll, python36.dll, libpython36m.dll, libpython3.6.dll, libpython3.6m.dll
    This would mean your Python installation doesn't come with proper library files.
    This usually happens by missing development package, or unsuitable build parameters of Python installation.

    * On Debian/Ubuntu, you would need to install Python development packages
      * apt-get install python3-dev
      * apt-get install python-dev
    * If you're building Python by yourself, please rebuild your Python with `--enable-shared` (or, `--enable-framework` on Darwin)
```
To fix this, the DLL/SO finding needs an edit. So if you want to use PyInsatller from inside a PyInstaller executable, you'll need to submit a PR that changes how PyInstaller searches for the DLL/SO.

**Comment by bjones1** [comment]
This issue makes no sense to me: why freeze a frozen application? This can't be done -- a frozen application's layout can't be analyzed for freezing, since it's no longer a pure Python application.

**Comment by Legorooj** [comment]
@bjones1 I believe @masoudr is using PyInstaller to build another application from within a frozen application? If so I'll reopen.

**Comment by bjones1** [comment]
Correct, which is why it can't work. PyInstaller has to analyzer Python source, not frozen, re-arranged frozen application stuff. I doubt ModuleGraph will understand how to peek inside the .exe file to find an archive of compressed source files; hook assume they're operating on the unfrozen source, and will be confused; etc.

**Comment by Legorooj** [comment]
>@bjones1 I believe @masoudr is using PyInstaller to build another application from within a frozen application? If so I'll reopen.

You misunderstand. (I think) @masoudr is building app A, which uses PyInstaller internally. Then, he runs app A, which in turn compiles app B. @masoudr am I correct?

**Comment by masoudr** [comment]
@Legorooj @bjones1 Let me clear this for you.
First I'm not doing a non-sense work, if I had any other option I would definitely do that.
Suppose that I have an application like App A. In this app, I have a script that accepts some external executable that needs to be provided by the user to work properly. After that App A should be frozen and used for the final executable.
To automate this, I created a GUI app (App B) that accepts this executable and uses PyInstaller to freeze App A. Here I used the PyInstaller API.
This app works fine without any problem. The problem here is that App B is a python script and it needs to become frozen. Currently, I solved the problem by using PyInstaller executable instead of its API in App B. But this forces me to add a separate python environment which is not ideal.
